### ISSUE TITLE
Provide ATC (REST) API docs

### ISSUE THREAD

**Issue by vito** [concourse/concourse#1122]
---
Moved from concourse/atc#80
/cc @bechtoldt

**Comment by trobert2** [comment]
any progress on this issue?
inb4 "nobody likes writing docs"

**Comment by scarytom** [comment]
I guess this is the closest thing we have to documentation for the API: https://github.com/concourse/concourse/blob/master/atc/routes.go#L106

**Comment by jama-pivotal** [comment]
Whelp, I know its been a while but we're actually re-visiting this now. Since we originally wrote the issue, we've held off from doing this in favour of implementing Users and RBAC.

We've finished both of those features now so  we're starting to revisit what it will take to build up a sustainably maintained and documented API. As a first step, we're asking for feedback from any Concourse users who have written any scripts/apps/code against the existing Concourse API. If you have any feedback, please comment below!

**Comment by ifraixedes** [comment]
An Open API document would be highly appreciated. It will allow knowing the request and response payloads on each route without having to dig into the Go source code.

On the other hand (and probably in a second stage) some more source comments (GoDoc) would also be appreciated.

**Comment by murac** [comment]
Even just for the time being, an example of fetching a Bearer token via script would be great. Anyone have any leads there?

**Comment by StevenArmstrong** [comment]
Example to trigger a job and fetch a bearer token in process

#!/usr/bin/python3
               import requests
               import json
               import sys
               concourse_url = "{{ concourse_url }}"
               username = "{{ concourse_user }}"
               password = "{{ concourse_password }}"
               pipeline = "{{ feature }}"
               job = "deploy feature in {{ env }}"
               url = concourse_url + "/sky/token"
               req_sessions = requests.Session() #load session instance
               req_sessions.auth = ('fly', 'Zmx5')
               headers = {'Content-Type':
'application/x-www-form-urlencoded'}
               data = {
                   'grant_type': 'password',
                   'password': password,
                   'scope': 'openid profile email federated:id groups',
                   'username': username
               }
               r = req_sessions.post(url, headers=headers, data=data,
verify=False)
               print(r.text)
               json_data = json.loads(r.text)
               print(json_data["access_token"])
               req_session = requests.Session() #load session instance
               headers = {'Authorization': 'Bearer '+
json_data["access_token"]}
               req = req_session.post(concourse_url +
"/api/v1/teams/main/pipelines/" + pipeline + "/jobs/" + job + "/builds",
headers=headers, verify=False)
               print(req.content)

**Comment by dwelch2344** [comment]
It works like a charm. Syntax was a little hard to understand in the dogs when I first read through. Might open a PR for a little more clarity for newbies like me. Would that be received well? Either way, thanks for the awesome library!

**Comment by scarytom** [comment]
Could someone help me out with connecting via curl please?  I'm trying to replicate @StevenArmstrong 's python example, but the token request keeps giving me a 401.  Here is what I have:

```sh
curl -v \
-X POST \
-H 'Content-Type: application/x-www-form-urlencoded' \
--user 'sky:Zmx5' \
--data-urlencode 'username=foo' \
--data-urlencode 'password=bar' \
--data-urlencode 'grant_type=password' \
--data-urlencode 'scope=openid profile email federated:id groups' \
'https://my-concourse.com/sky/token'
```

**Comment by vergissberlin** [comment]
An API documentation is created very easily with swagger. You can also generate code with it.

**Comment by BlacktoBlue** [comment]
@scarytom in that post what does is user and scope?

**Comment by predatorian3** [comment]
@scarytom , could you explain or elaborate on your cURL example? Why use two username and passwords, and what is the scope coming from?

**Comment by scarytom** [comment]
@BlacktoBlue @predatorian3 You should ask these questions of @StevenArmstrong, as my curl is little more than a re-work of his python example.

As far as I understand it, the ''fly', 'Zmx5`` combo is simply a hard-coded basic http auth inside all concourse deployments -- no idea if that's configurable or not.  The secure login cones from the other pair in the POST form.  I can't tell you much about scope.

**Comment by StevenArmstrong** [comment]
Yeah we wiresharked the requests and went through the code to see what was
required to make the call. The python script works above. Although rendered
bad on non email and mobile.

**Comment by SvenDowideit** [comment]
@dwelch2344 thought I'd play around a little with postman - yup, its kinda fun

https://sven-test.postman.co/collections/6836899-9f94750c-9fc3-4706-8754-d9f22b039bb1?workspace=360d60e5-9d57-4b24-8b6f-69017dcbc747#f7f6a2e6-04f4-45f7-8a61-08ac852790de

https://app.getpostman.com/join-team?invite_code=252685ea862969ea8b0fdab2309e73ec

**Comment by SvenDowideit** [comment]
for those that come later - its also excellent to know that `fly` has a `--verbose` flag that will show you the HTTP request and responses that it makes - ie, the client is self documenting, the only hard thing is discovering that it does it :)

```
--verbose              Print API requests and responses
```

**Comment by iornstein** [comment]
@SvenDowideit (or anyone else) how do we add the authorization to requests like fetching the old config or PUT-ing new config for pipelines? I hit these end points and get 401. For example

`GET {{local-concourse-url}}/api/v1/teams/main/pipelines/hello-world/config`

Results in a 401. What is the authorization I am supposed to add?

**Comment by fewknow** [comment]
Fetching bearer token with basic auth(local)

#!/bin/bash

## Variables required  #need to update these to take inputs for getting token per team and target.
CONCOURSE_URL="http://concourseci.XXXXXX.XXX"
CONCOURSE_USER="XXXX"
CONCOURSE_PASSWORD="XXXX"
CONCOURSE_TEAM="XXXXXX"
CONCOURSE_TARGET=XXXXXX

function get_token() {
  ## Create a file named token that will be used to read and write tokens
  touch token

  ## extract the LDAP authentication url and write to token file
  LOCAL_AUTH_URL=$CONCOURSE_URL$(curl -b token -c token -L "$CONCOURSE_URL/sky/login" -s | grep "/sky/issuer/auth/local?" | awk -F'"' '{print $4}')
  echo "url is $LOCAL_AUTH_URL"
  # login using username and password while writing to the token file
  curl -s -o /dev/null -b token -c token  -L --data-urlencode "login=$CONCOURSE_USER" --data-urlencode "password=$CONCOURSE_PASSWORD" "$LOCAL_AUTH_URL"

  ATC_BEARER_TOKEN=`grep 'Bearer' token | cut -d\   -f2 | sed 's/"$//'`
  echo $ATC_BEARER_TOKEN
}


Here is a link for LDAP : https://github.com/concourse/concourse/issues/1487

**Comment by TheDevMinerTV** [comment]
Bump

**Comment by aalfansuri** [comment]
any further progress on this?

**Comment by kalrashubham49** [comment]
> Here is a bearer token request using curl
> 
> ```shell
> curl -v \
> -X POST \
> -H 'Content-Type: application/x-www-form-urlencoded' \
> --user 'fly:Zmx5' \
> --data-urlencode 'username=foo' \
> --data-urlencode 'password=bar' \
> --data-urlencode 'grant_type=password' \
> --data-urlencode 'scope=openid profile email federated:id groups' \
> 'https://my-concourse.com/sky/token'
> ```

It isn't working in my case , can you please verify does it still work in concourse 6.6.0

**Comment by kalrashubham49** [comment]
> Even just for the time being, an example of fetching a Bearer token via script would be great. Anyone have any leads there?

Did you get the script for this ?

**Comment by scarytom** [comment]
> > Here is a bearer token request using curl
> > ```shell
> > curl -v \
> > -X POST \
> > -H 'Content-Type: application/x-www-form-urlencoded' \
> > --user 'fly:Zmx5' \
> > --data-urlencode 'username=foo' \
> > --data-urlencode 'password=bar' \
> > --data-urlencode 'grant_type=password' \
> > --data-urlencode 'scope=openid profile email federated:id groups' \
> > 'https://my-concourse.com/sky/token'
> > ```
> 
> It isn't working in my case , can you please verify does it still work in concourse 6.6.0

I have not verified this on any recent concourse versions, sorry.

**Comment by kalrashubham49** [comment]
> for those that come later - its also excellent to know that `fly` has a `--verbose` flag that will show you the HTTP request and responses that it makes - ie, the client is self documenting, the only hard thing is discovering that it does it :)
> 
> ```
> --verbose              Print API requests and responses
> ```
Is Verbose flag working ??

**Comment by micheelengronne** [comment]
Nope. It shows the response but not the request.

**Comment by gkspranger** [comment]
@kalrashubham49 @micheelengronne 

the URL has changed in later versions -- don't know when exactly

as of 6.7.2 -- it is `/sky/issuer/token`

found it by searching here :: https://github.com/concourse/concourse/search?q=Zmx5

**Comment by jlamb1** [comment]
@fewknow 's script worked great! In case anyone else makes the silly mistake I did, the `--user 'fly:Zmx5'` shouldn't be modified, but entered as is.

**Comment by fedoranvar** [comment]
postman's way

![image](https://user-images.githubusercontent.com/52028706/203256092-3aaf5691-770d-4c0e-82af-93f9aaa759e0.png)

![image](https://user-images.githubusercontent.com/52028706/203256199-231ac13b-990e-4d7e-bc57-c8900bc93785.png)

**Comment by taylorsilva** [comment]
Would still be nice to have. We should probably update the website since I think it's a bit hard for people to grok how to add things to it beyond basic text stuff.

**Comment by IvanChalukov** [comment]
Hi all - @Kump3r and I have prepared a draft of the Concourse API documentation in Markdown format. We hope it’s useful for anyone interested in working with or extending the ATC REST API. You can find the draft in the [API Endpoint Quick Reference
](https://github.com/IvanChalukov/concourse-api-docs?tab=readme-ov-file#-api-endpoint-quick-reference) section in [concourse-api-docs](https://github.com/IvanChalukov/concourse-api-docs) repo.

@taylorsilva Since the official documentation is currently being migrated to MCdocs (as mentioned in the last cff working group meeting), we weren’t sure where contributions like this should live. Once the migration is complete, we’re happy to open a PR to integrate the API docs into the new structure.

If you have a preferred location or any guidance on how you’d like this incorporated, just let us know - we’re glad to help.

**Comment by taylorsilva** [comment]
I think having it sit in a separate repo is a good spot for now. Thanks for sharing what you've got so far!

The website migration will be held off until we can migrate directly to zensical, which looks like it won't be too hard.

Zensical mentions they'll have integrated support for mkdocstrings: https://zensical.org/compatibility/plugins/
Which may make maintaining the API docs easier too!

Either way, this repo looks great. Once we've migrated the site, we can do a PR of your repo's content into the concourse/docs repo.
